subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/ImmutableTransformationWorkspaceProvider.java

```diff
@@ -17,47 +17,55 @@
 import org.gradle.cache.internal.SingleDepthFilesFinder;
 import org.gradle.internal.Try;
 import org.gradle.internal.execution.history.ExecutionHistoryStore;
+import org.gradle.internal.execution.history.impl.DefaultExecutionHistoryStore;
 import org.gradle.internal.file.FileAccessTimeJournal;
 import org.gradle.internal.file.impl.SingleDepthFileAccessTracker;
 
 import javax.annotation.concurrent.NotThreadSafe;
 import java.io.Closeable;
 import java.io.File;
 
-import static org.gradle.api.internal.artifacts.ivyservice.CacheLayout.TRANSFORMS_STORE;
 import static org.gradle.cache.internal.LeastRecentlyUsedCacheCleanup.DEFAULT_MAX_AGE_IN_DAYS_FOR_RECREATABLE_CACHE_ENTRIES;
 import static org.gradle.cache.internal.filelock.LockOptionsBuilder.mode;
 
 @NotThreadSafe
 public class ImmutableTransformationWorkspaceProvider implements TransformationWorkspaceProvider, Closeable {
-    private static final int FILE_TREE_DEPTH_TO_TRACK_AND_CLEANUP = 1;
+    private static final int FILE_TREE_DEPTH_TO_TRACK_AND_CLEANUP = 2;
 
     private final SingleDepthFileAccessTracker fileAccessTracker;
-    private final File filesOutputDirectory;
+    private final File baseDirectory;
     private final ExecutionHistoryStore executionHistoryStore;
     private final PersistentCache cache;
 
-    public ImmutableTransformationWorkspaceProvider(File transformsStoreDirectory, CacheRepository cacheRepository, FileAccessTimeJournal fileAccessTimeJournal, ExecutionHistoryStore executionHistoryStore) {
-        filesOutputDirectory = new File(transformsStoreDirectory, TRANSFORMS_STORE.getKey());
-        this.executionHistoryStore = executionHistoryStore;
-        cache = cacheRepository
-            .cache(transformsStoreDirectory)
-            .withCleanup(createCleanupAction(filesOutputDirectory, fileAccessTimeJournal))
+    public ImmutableTransformationWorkspaceProvider(
+        File baseDirectory,
+        CacheRepository cacheRepository,
+        FileAccessTimeJournal fileAccessTimeJournal,
+        InMemoryCacheDecoratorFactory inMemoryCacheDecoratorFactory,
+        StringInterner stringInterner
+    ) {
+        this.baseDirectory = baseDirectory;
+        this.cache = cacheRepository
+            .cache(baseDirectory)
+            .withCleanup(createCleanupAction(baseDirectory, fileAccessTimeJournal))
             .withCrossVersionCache(CacheBuilder.LockTarget.DefaultTarget)
             .withDisplayName("Artifact transforms cache")
             .withLockOptions(mode(FileLockManager.LockMode.OnDemand)) // Lock on demand
             .open();
-        fileAccessTracker = new SingleDepthFileAccessTracker(fileAccessTimeJournal, filesOutputDirectory, FILE_TREE_DEPTH_TO_TRACK_AND_CLEANUP);
+        this.fileAccessTracker = new SingleDepthFileAccessTracker(fileAccessTimeJournal, baseDirectory, FILE_TREE_DEPTH_TO_TRACK_AND_CLEANUP);
+        this.executionHistoryStore = new DefaultExecutionHistoryStore(() -> cache, inMemoryCacheDecoratorFactory, stringInterner);
```

Comment:
The execution history must not be part of the cross version cache directory, since it is version specific. So it needs to go back into the version specific caches directory.

Reply:
Hm, you're right. However, what's the point of having the transforms cache be global if the IDs include the Gradle version anyway? Shouldn't the whole thing be version-specific, now that we are changing the location anyway?

Reply:
The same reason can be applied for the build cache, since everything in there also contains the Gradle version. So I am not sure why we don't use a version specific directory for the transforms, I suppose because we didn't use to.

Reply:
Is it about cleanup? So this way stuff used by old versions also gets cleared up eventually?

Reply:
I reverted that change for now.